source/common/upstream/health_checker_impl.cc

```diff
@@ -345,7 +385,7 @@ bool HttpHealthCheckerImpl::HttpActiveHealthCheckSession::shouldClose() const {
     return false;
   }
 
-  if (!parent_.reuse_connection_) {
+  if (!parent_.reuse_connection_ || received_no_error_goaway_) {
```

Comment:
Curious, could we just set parent_.reuse_connection_ false when get the goaway, and let all the rest of the code proceed as usual?  I assume if we're not reusing the connection it would close on stream reset or stream complete.

Reply:
i think we need both because otherwise we wouldn't know what to reset `parent_.reuse_connection_` to when we start a new stream

Reply:
Maybe this is what you meant: one thing I could do is name this variable `reuse_connection_`, and at the start of the stream set it to `parent.reuse_connection_`, and then in set it to false if there is a NO_ERROR GOAWAY. That might make it more clear